src/main/scala/org/bitcoins/core/script/crypto/HashType.scala

```diff
@@ -1,47 +1,145 @@
 package org.bitcoins.core.script.crypto
 
-import org.bitcoins.core.util.{BitcoinSUtil}
+import org.bitcoins.core.number.Int32
+import org.bitcoins.core.script.crypto.SIGHASH_ALL.SIGHASH_ALLImpl
+import org.bitcoins.core.util.{Factory, BitcoinSUtil}
 
 /**
  * Created by chris on 1/18/16.
  */
 sealed trait HashType {
-  def byte : Byte
-  def hex : String = BitcoinSUtil.encodeHex(byte)
+  type HashType
+  def num : Int32
 }
 
+trait HashTypeOperations extends Factory[HashType] {
+  def fromBytes(bytes : Seq[Byte]) : HashType = {
+    val num = Int32(bytes)
+    fromNumber(num)
+  }
+
+  def fromNumber(num : Int32) : HashType = {
+    if (isSIGHASH_NONE(num)) {
+      if (isSIGHASH_NONE_ANYONECANPAY(num)) SIGHASH_NONE_ANYONECANPAY(num) else SIGHASH_NONE(num)
+    }
+    else if (isSIGHASH_SINGLE(num)) {
+      if (isSIGHASH_SINGLE_ANYONECANPAY(num)) SIGHASH_SINGLE_ANYONECANPAY(num) else SIGHASH_SINGLE(num)
+    }
+    else if (isSIGHASH_ANYONECANPAY(num)) {
+      if (isSIGHASH_ALL_ANYONECANPAY(num)) SIGHASH_ALL_ANYONECANPAY(num) else SIGHASH_ANYONECANPAY(num)
+    }
+    else {
+      SIGHASH_ALL(num)
+    }
+  }
+
+  def byte (hashType : HashType) : Byte = hashType match {
+    case zero if zero.num == Int32.zero => 0x00.toByte
```

Comment:
Are these just the same numbers that are in the companion object? I.e. could we have `SIGHASH_SINGLE.value` instead of just `0x01.toByte`?